Check that the Docker API responds before trying to use it.

UnaKarlsen opened the pull request:
As mentioned in issue #13 the log collector will sometimes hang when querying Docker for information. This is usually the result of the Docker Daemon hanging and the Docker client does not time out when making requests.

To work around this I have split the data collection tasks off that need the Docker API to be responding vs the ones that don't. We then check if docker is running. If it is, we ping the Docker API, using a timeout of 60 seconds. If the API responds OK, then we can query the API for information. If not, inform the user that some information will be missing.

The important thing here is that data is still collected and packed, even if some tasks do not work, providing at least some of the info needed to diagnose issues.

UnaKarlsen committed:
Added is_docker_healthy which checks if the Docker Daemon is running, then checks that it's API is responding. Only if these pass, then collect the data which requires the Docker API to be healthy.

UnaKarlsen committed:
Changed is_docker_healthy() and get_containers_info() to provide a warning() if they fail rather than die(). This allows at least some information to be collected, rather than nothing.

UnaKarlsen committed:
Seperated the ECS Agent info collection out of get_containers_info() and put it in get_agent_info(). This allows easier seperatation of tasks that need the Docker API and those that don't.

UnaKarlsen committed:
Some ECS Agent info can be collected even if the agent is not running, so I have moved them out of the block that only runs if the agent is running.

UnaKarlsen committed:
Fixed is_docker_healthy() to return 1 if the Docker Daemon is not running and 2 if the API is not responding.

UnaKarlsen committed:
Added check that the result from the _ping command is OK.

UnaKarlsen committed:
Fixed typo in if statement and silenced unnesisary verbosity in the curl statement in is_docker_healthy()

jhaynes commented on ecs-logs-collector.sh:
```diff
@@ -395,64 +400,101 @@ get_system_services()
-
     ok
 
+    if [ -e /usr/bin/curl ]; then
```
Why use the absolute path here instead of checking that `curl` is on the users' path?

jhaynes commented on ecs-logs-collector.sh:
```diff
@@ -395,64 +400,101 @@ get_system_services()
 {
-  try "gather Docker daemon information"
-
+  try "Checking if Docker is running"
```
This will read oddly since the `try` function prepends the string with `Trying to `.  maybe `try "confirm that docker is running"`

jhaynes commented on ecs-logs-collector.sh:
```diff
@@ -395,64 +400,101 @@ get_system_services()
     ok
 
+    if [ -e /usr/bin/curl ]; then
+      try "Checking if Docker API is responding"
```
Same as above. Something like: `try "get a response from the Docker API"`

jhaynes commented on ecs-logs-collector.sh:
```diff
@@ -395,64 +400,101 @@ get_system_services()
+        else
+          warning "The Docker API did not respond with OK. Some info will be unavailable."
+          echo "API output was:"
+          echo $result
```
Nit: I might condense these three into a single line. `warning "The Docker API responded with $result. Some info  will be unavailable."

jhaynes commented on ecs-logs-collector.sh:
```diff
@@ -395,64 +400,101 @@ get_system_services()
+          warning "The Docker API did not respond with OK. Some info will be unavailable."
+          echo "API output was:"
+          echo $result
+          return 3
```
Since the function name `is_docker_healthy` sounds like it returns a boolean, yes or no, I'm tempted to say this should only return 1 or 0. However, if we want more specific information from it that we might act on later, see http://tldp.org/LDP/abs/html/exitcodes.html for some exit code suggestions. I might use `1` for this and `126` for the failed curl.

jhaynes reviewed (commented):
Thanks for adding this. It will go a long way to improving the user experience with this tool.

We might also want to add timeouts to our individual Docker calls that might hang. something like `timeout 10s docker ...`. It will exit 124 if the timeout is hit and we can handle that and inform the user appropriately. 

I'd implement it as a wrapper function, e.g. `docker_with_timeout`, that accepts the long docker command and does all the handling and timing out of the passed command inside.

UnaKarlsen committed:
Changed the wording of Try entries
Changed to not use the absolue path for curl.
Changed the return code for is_docker_healthy

SmilyBorg commented:
I have made updates to fix the wording of messages to the user, removed the need for an absolute path for curl and updated the return codes in is_docker_healthy().

Using timeout sounds good, though if Docker does not respond to it's _ping API call, it's not going to respond to any other API call so I'm not sure if it really adds anything to the user experience. It mainly just slows things down a lot.

That said, adding timeout to each docker command, along with the _ping API test makes sense in the event that Docker stops responding correctly after the Ping succeeds. I'll create a wrapper as you suggested and include that soon.

jhaynes commented on ecs-logs-collector.sh:
```diff
@@ -395,64 +400,99 @@ get_system_services()
-
     ok
 
+    if [ `which curl` ]; then
```
If curl isn't installed, this will fail silently. Is that intended?

UnaKarlsen replied:
That wasn't intended. I have now added a warning message if curl is not found, but I still allow collection of information in the usual way since we don't know if the API is working or not.

jhaynes commented on ecs-logs-collector.sh:
```diff
@@ -395,64 +400,99 @@ get_system_services()
+  pgrep agent > /dev/null
+  if [[ "$?" -eq 0 ]]; then
+
+    if [ -e /usr/bin/curl ]; then
```
should be `which curl` in case curl is installed but not at `/usr/bin`. Or just `curl` since if it is installed, it will almost certainly be on the path.

UnaKarlsen committed:
Added error messages so that the error from curl being missing is not swallowed.
Removed further reference to curl by path

jahkeup commented:
Not all distros listed on our supported list include a `curl` package with the `--unix-socket` flag. The flag is [relatively new (added in 7.40)](https://curl.haxx.se/docs/releases.html). 

* Amazon Linux: Supported (7.53.1)
* Ubuntu 14.04:  [Not Supported (7.35.0)](https://packages.ubuntu.com/trusty/curl)
* Debian 8: [Not Supported (7.38.0)](https://packages.debian.org/jessie/curl)
* RHEL 7: Not Supported (7.29)

jhaynes commented:
Instead of trying the socket, we could just run the command we want with `timeout`.

jahkeup commented:
This was addressed with changes in #32, I'm going to close this PR out. Please feel free to make and contribute additional improvements with another PR if #32 doesn't accommodate your scenario.

Thanks!
